extensions/amp-a4a/0.1/a4a-variable-source.js

```diff
@@ -113,10 +112,68 @@ export class A4AVariableSource extends VariableSource {
       return getNavigationData(this.win_, 'redirectCount');
     });
 
+    this.set('HTML_ATTR',
+        /** @type {function(...*)} */(this.htmlAttrBinding_.bind(this)));
+
     for (let v = 0; v < WHITELISTED_VARIABLES.length; v++) {
       const varName = WHITELISTED_VARIABLES[v];
       const resolvers = this.globalVariableSource_.get(varName);
       this.set(varName, resolvers.sync).setAsync(varName, resolvers.async);
     }
   }
+
+  /**
+   * Provides a binding for getting attributes from the DOM.
+   * Most such bindings are provided in src/service/url-replacements-impl, but
+   * this one needs access to this.win_.document, which if the amp-analytics
+   * tag is contained within an amp-ad tag will NOT be the parent/publisher
+   * page. Hence the need to put it here.
+   * @param {string} cssSelector Elements matching this selector will be
+   *     included, provided they have at least one of the attributeNames
+   *     set, up to a max of 10. May be URI encoded.
+   * Note: Additional params will be the names of the attributes whose
+   *     values will be returned. There should be at least 1.
+   * @returns {string} A stringified JSON array containing one member for each
+   *     matching element. Each member will contain the names and values of the
+   *     specified attributes, if the corresponding element has that attribute.
+   *     Note that if an element matches the cssSelected but has none of the
+   *     requested attributes, then nothing will be included in the array
+   *     for that element.
+   */
+  htmlAttrBinding_(cssSelector) {
+    const HTML_ATTR_MAX_ELEMENTS = 10;
+    const HTML_ATTR_MAX_ATTRS = 10;
+    const attributeNames = Array.prototype.slice.call(arguments, 1);
+    if (!cssSelector || !attributeNames.length) {
+      return '[]';
+    }
+    cssSelector = decodeURI(cssSelector);
+    let elements;
+    try {
+      elements = this.win_.document.querySelectorAll(cssSelector);
+    } catch (e) {
+      const TAG = 'A4AVariableSource';
+      user().error(TAG, `Invalid selector: ${cssSelector}`);
+      return '[]';
+    }
+    const result = [];
+    for (let i = 0; i < elements.length &&
+        result.length < HTML_ATTR_MAX_ELEMENTS; ++i) {
```

Comment:
I see what you mean. I propose having an additional limit:
* HTML_ATTR_MAX_ELEMENTS can be renamed to HTML_ATTR_MAX_ELEMENTS_TO_RETURN
* We can introduce a new constant, something like HTML_ATTR_MAX_ELEMENTS_TO_TRAVERSE, and set this to something a bit higher like 20 or 50. If querySelectorAll returns more than this number, then return an error. Otherwise proceed to inspect attributes, but only until we find HTML_ATTR_MAX_ELEMENTS_TO_RETURN matching elements.

Reply:
Can we set `HTML_ATTR_MAX_ELEMENTS_TO_TRAVERSE ` limit to 20 instead : ) I really don't think anyone will want to query sooooo much elements. Given all they want is image url.